### PR TITLE
[HttpClient] Allow to pass user/pw as an array

### PR DESCRIPTION
**Pull request by dunglas**
| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | n/a   
| License       | MIT
| Doc PR        | n/a

Makes the API a bit more "PHPish".

### CODE REVIEW

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/HttpClient/HttpOptions.php [hunk 1]
```diff
@@ -34,9 +34,12 @@ public function toArray(): array
     /**
      * @return $this
      */
-    public function setAuth(string $userinfo)
+    public function setAuth(string $user, string $password = '')
     {
-        $this->options['auth'] = $userinfo;
+        $this->options['auth'] = $user;
+        if ('' === $password) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/HttpOptions.php]: !==

File: src/Symfony/Component/HttpClient/HttpClientTrait.php [hunk 2]
```diff
@@ -71,8 +71,17 @@ private static function prepareRequest(?string $method, ?string $url, array $opt
             throw new InvalidArgumentException(sprintf('Option "on_progress" must be callable, %s given.', \is_object($onProgress) ? \get_class($onProgress) : \gettype($onProgress)));
         }
 
+        if (\is_array($options['auth'] ?? null)) {
+            $count = \count($options['auth']);
+            if ($count <= 0 || $count > 2) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/HttpClientTrait.php]: `if (!\in_array(\count($options['auth']), [1, 2], true))`?

Comment (dunglas) [reply src/Symfony/Component/HttpClient/HttpClientTrait.php]: As you wish!

Comment (dunglas) [reply src/Symfony/Component/HttpClient/HttpClientTrait.php]: changed

Comment (fabpot) [reply src/Symfony/Component/HttpClient/HttpClientTrait.php]: TBH, I prefer the more explicit way.

**Comment by fabpot** [comment]
This one needs to be rebased :)

File: src/Symfony/Component/HttpClient/HttpClientTrait.php (hunk 2 above)
Comment (dunglas) [reply src/Symfony/Component/HttpClient/HttpClientTrait.php]: reverted :)

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @dunglas.

**Comment by javiereguiluz** [comment]
Do we really want to allow two different ways of doing the same thing? Maybe allowing only the use of arrays instead of strings could be enough?

**Comment by fabpot** [comment]
I wouldn't mind having only the array way. Makes sense to me. WDYT @dunglas @nicolas-grekas?

**Comment by dunglas** [comment]
Being allowed to pass a string doesn't hurt IMO (as it's a standard), but if we have to choose only one I'll go for the array way too.

**Comment by nicolas-grekas** [comment]
The string is the only thing that the standard describes. The array is only sugar, and leaks actually (you can't use a `:` is the user, per the standard, and unlike the array may hint you).
The array also doesn't play well with env vars.

I'm for keeping the string of there should be only one.

**Pull request merged**